go/adbc/driver/snowflake/connection.go

```diff
@@ -557,13 +587,204 @@ func (c *cnxn) populateMetadata(ctx context.Context, depth adbc.ObjectDepth, cat
 	return metadataRecords, nil
 }
 
+func (c *cnxn) populateConstraintSchema(ctx context.Context, depth adbc.ObjectDepth, metadataRecords []internal.Metadata) (map[internal.CatalogSchemaTable][]internal.ConstraintSchema, error) {
+	constraintLookup := make(map[internal.CatalogSchemaTable][]internal.ConstraintSchema)
+	tableConstraintsData, err := c.getConstraintsData(ctx, depth, metadataRecords)
+	if err != nil {
+		return nil, err
+	}
+
+	fullyQualifiedConstraintSchemaLookup := make(map[string]internal.ConstraintSchema)
+	for _, data := range tableConstraintsData {
+		var fullyQualifiedConstraintName string
+		if data.fk_constraintName.Valid {
+			fullyQualifiedConstraintName = getFullyQualifiedConstraintName(data.fkDbName.String, data.fkSchema.String, data.fkTblName.String, data.fk_constraintName.String)
+			if _, exists := fullyQualifiedConstraintSchemaLookup[fullyQualifiedConstraintName]; !exists {
+				fullyQualifiedConstraintSchemaLookup[fullyQualifiedConstraintName] = internal.ConstraintSchema{
+					ConstraintName:        data.fk_constraintName.String,
+					ConstraintType:        data.constraintType,
+					ConstraintColumnNames: []string{data.fkColName.String},
+					ConstraintColumnUsages: []internal.UsageSchema{
+						{
+							ForeignKeyCatalog:  data.dbName,
+							ForeignKeyDbSchema: data.schema,
+							ForeignKeyTable:    data.tblName,
+							ForeignKeyColName:  data.colName,
+						},
+					},
+				}
+			} else {
+				constraintInfo := fullyQualifiedConstraintSchemaLookup[fullyQualifiedConstraintName]
+				constraintInfo.ConstraintColumnNames = append(constraintInfo.ConstraintColumnNames, data.fkColName.String)
+
+				constraintInfo.ConstraintColumnUsages = append(constraintInfo.ConstraintColumnUsages, internal.UsageSchema{
```

Comment:
Any way to reduce duplication here vs at line 609?

Reply:
since both cases use this `UsageSchema` you should be able to populate it before line 602 and then use the populated schema in both places rather than duplicating it

Reply:
refactored into a separate method.